File: kaop/src/main/java/io/github/coyamo/kaop/Pointcut.java
package io.github.coyamo.kaop;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.reflect.Method;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.coyamo.kaop.annotation.AopJoinPoint;

/**
 * @author cxc
 * @date 2022/11/9 17:13
 * @version 1.0
 */
public class Pointcut {
    final Object owner;
    private final Map<String, Method> proceedMap;
    final Map<String, List<Class<? extends Aspect>>> pluginMap;

    final Map<String, AbsJoinPoint> cache = new HashMap<>();

    public Pointcut(@NonNull Object owner, @NonNull Map<String, Method> proceedMap, @NonNull Map<String, List<Class<? extends Aspect>>> pluginMap) {
        this.owner = owner;
        this.proceedMap = proceedMap;
        this.pluginMap = pluginMap;
    }



    @Nullable
    public JoinPoint joinPoint(String point) {
        AbsJoinPoint joinPoint = cache.get(point);
        if(joinPoint instanceof JoinPoint){
            return (JoinPoint)joinPoint;
        }
        Method method = proceedMap.get(point);
        if (method != null) {
            List<Class<? extends Aspect>> list = pluginMap.get(point);
            joinPoint = new JoinPoint(owner, method, list == null ? Collections.emptyList() : list);
            cache.put(point, joinPoint);
            return (JoinPoint)joinPoint;
        }
        return null;
    }
}

File: kaop/src/main/java/io/github/coyamo/kaop/Pointcut.kt
package io.github.coyamo.kaop

import android.util.Log
import android.util.Pair
import io.github.coyamo.kaop.KAop.DEFAULT_ORDER
import io.github.coyamo.kaop.KAop.ORDER_FILED_NAME
import java.lang.reflect.Method

/**
 * @author Coyamo
 * @date 2022/12/17 00:47
 * @version 1.0
 */
class Pointcut(
    private val owner: Any
) {
    private val pluginMap = mutableMapOf<String, List<Class<out Aspect?>>>()
    private val cache: MutableMap<String, AbsJoinPoint> = HashMap()

    inline operator fun <reified T> invoke(noinline block: PointScope.() -> T): T {
        //这里必须要新建一个对象才可以获取到enclosingMethod
        return object :MethodGetter<T>(this, block){}.proxy()
    }

    /**
     * 和[invoke]作用一样，java代码使用
     */
    fun <T> pointcut(methodGetter:MethodGetter<T>, block: PointScope.() -> T):T{
        val method = methodGetter.javaClass.enclosingMethod
        val scope = PointScope()
        return if (method == null) {
            Log.w("KAop", "Obtaining enclosing method failed!")
            scope.block()
        } else {
            val joinPoint = joinPoint(method, scope, block)
            return if (joinPoint == null) {
                scope.block()
            } else {
                joinPoint.join() as T
            }
        }
    }


    private fun joinPoint(method: Method, scope: PointScope, block: PointScope.() -> Any?): JoinPoint? {
        val point = method.toString()
        var joinPoint = cache[point]
        //默认名字是否有缓存
        if (joinPoint is JoinPoint && joinPoint.method == method) {
            joinPoint.scope = scope
            joinPoint.block = block
            Log.d("KAop", "Load ${method.name} from cache")
            return joinPoint
        }
        val annotations = method.annotations

        var isPointcut = false
        val tmpPlugins: MutableList<Pair<Int, Class<out Aspect?>>> = ArrayList()
        for (annotation in annotations) {
            val annotationClass: Class<*> = annotation.annotationClass.java
            val aopPluginAnnotation = annotationClass.getAnnotation(
                AspectAnnotation::class.java
            )
            if (aopPluginAnnotation != null) {
                isPointcut = true
                var order = DEFAULT_ORDER
                try {
                    val orderMethod = annotation.javaClass.getDeclaredMethod(ORDER_FILED_NAME)
                    val returnType = orderMethod.returnType
                    if (returnType == Int::class.java || returnType == Int::class.javaPrimitiveType) {
                        order = orderMethod.invoke(annotation) as Int
                    }
                } catch (ignore: Exception) {
                }
                tmpPlugins.add(Pair(order, aopPluginAnnotation.plugin.java))
            }
        }

        if (!isPointcut) {
            Log.d("KAop", "${method.name}:Redundant operations")
            return null
        }

        //排序
        tmpPlugins.sortWith { o1, o2 ->
            Integer.compare(
                o1.first,
                o2.first
            )
        }
        val plugins: MutableList<Class<out Aspect?>> = ArrayList()
        for (tmpPlugin in tmpPlugins) {
            plugins.add(tmpPlugin.second)
        }
        pluginMap[point] = plugins
        joinPoint = JoinPoint(owner, method, scope, block, plugins)
        cache[point] = joinPoint
        return joinPoint
    }
}